### ISSUE TITLE
Kerberos issue - Can't find client principal *** in cache collection

### ISSUE THREAD

**Issue by AseemKumar** [ansible/ansible#17285]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

kerberos (1.2.4), pykerberos (1.1.13), pycrypto (2.6.1), pywinrm (0.2.0)
##### ANSIBLE VERSION

```
ansible 2.1.1.0
```
##### CONFIGURATION

default config
##### OS / ENVIRONMENT

Ansible on Ubuntu, trying to manage windows servers
##### SUMMARY

Many a times, kerberos stops to work. I am able to get a ticket and a cache file gets generated but ansible output says ""msg": "kerberos: authGSSClientInit() failed: (('Unspecified GSS failure.  Minor code may provide more information', 851968), (\"Can't find client principal user@DOMAIN in cache collection\", -1765328243))","
This is an intermittent problem and have started seeing this only 4 days back, and no changes have been made. NTLM method works but Kerberos fails.
Have searched various forums and no luck, hence reaching out here. If this is not the right place, please guide me in the right direction.
##### STEPS TO REPRODUCE

Get a ticket from windows domain and then try to manage windows server connected to that domain.

```

```

##### EXPECTED RESULTS

The data in the cache file should have been picked up.
##### ACTUAL RESULTS

```
Loaded callback minimal of type stdout, v2.0
<servername> ESTABLISH WINRM CONNECTION FOR USER: user@domain on PORT 5986 TO servername
<servername> WINRM CONNECT: transport=kerberos endpoint=https://servername:5986/wsman
<servername> WINRM CONNECTION ERROR: authGSSClientInit() failed: (('Unspecified GSS failure.  Minor code may provide more information', 8
51968), ("Can't find client principal user@domain in cache collection", -1765328243))
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/ansible/plugins/connection/winrm.py", line 151, in _winrm_connect
    self.shell_id = protocol.open_shell(codepage=65001) # UTF-8
  File "/usr/local/lib/python2.7/dist-packages/winrm/protocol.py", line 132, in open_shell
    res = self.send_message(xmltodict.unparse(req))
  File "/usr/local/lib/python2.7/dist-packages/winrm/protocol.py", line 207, in send_message
    return self.transport.send_message(message)
  File "/usr/local/lib/python2.7/dist-packages/winrm/transport.py", line 170, in send_message
    prepared_request = self.session.prepare_request(request)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 394, in prepare_request
    hooks=merge_hooks(request.hooks, self.hooks),
  File "/usr/local/lib/python2.7/dist-packages/requests/models.py", line 299, in prepare
    self.prepare_auth(auth, url)
  File "/usr/local/lib/python2.7/dist-packages/requests/models.py", line 494, in prepare_auth
    r = auth(self)
  File "/usr/local/lib/python2.7/dist-packages/requests_kerberos/kerberos_.py", line 318, in __call__
    auth_header = self.generate_request_header(None, host, is_preemptive=True)
  File "/usr/local/lib/python2.7/dist-packages/requests_kerberos/kerberos_.py", line 158, in generate_request_header
    raise KerberosExchangeError("%s failed: %s" % (kerb_stage, str(error.args)))
KerberosExchangeError: authGSSClientInit() failed: (('Unspecified GSS failure.  Minor code may provide more information', 851968), ("Can't find client princ
ipal user@domain in cache collection", -1765328243))

servername | UNREACHABLE! => {
    "changed": false,
    "msg": "kerberos: authGSSClientInit() failed: (('Unspecified GSS failure.  Minor code may provide more information', 851968), (\"Can't find client princ
ipal user@domain in cache collection\", -1765328243))",
    "unreachable": true
}
```

**Comment by nitzmahone** [comment]
You shouldn't be using kerberos + pykerberos together, just pykerberos. Not sure if that's related or not, but I've not seen this behavior before. What do you have to do to recover?

**Comment by jhawkesworth** [comment]
When you get this, do a klist to see what you have in your ticket cache.
I have only used file based ticket cache - I believe there are keyring based kerb caches as well, so that's something to investigate.
When I have seen intermittent kerberos failures it has been because of clocks slipping or dns getting messed up due to duff DHCP behaviour (on mixed networks with fixed and wireless kit on the same network subdomain) - in both cases the error message was not the same though.

**Comment by AseemKumar** [comment]
@nitzmahone I am letting Ansible use whatever it wants to. The problem resolved by itself. I got on the kerberos IRC and the folks there suggested that I enable logging and change location of cache file. I changed the cache location and did kinit & kdestroy many a times and it started to work. I am unsure whether changing the cache location fixed the problem or something else.

@jhawkesworth I always do a klist and have always been getting a valid ticket. I tried connecting to different domains and also used another account and the same behaviour showed up every time. My clock was in sync and DNS was working fine.

I suspect it to be some odd issue with either pykerberos or the way ansible called it. I have no idea how to get more information to help you guys if this repeats.

**Comment by AseemKumar** [comment]
This problem is now coming up pretty often. Is there some kind of logs I can send that may help?

**Comment by nitzmahone** [comment]
Ansible and pykerberos both just use the configured libkrb5 default ccache (which location/type can be overridden by the usual envvars). There's nothing special going on there, and you're the first report I've seen of this behavior, so I'd be suspicious of some external process that's munging/locking/something else your ccache (kerb-using cron job, other software ?). Maybe the kerb folks can help out for further diagnostics, because we're just consuming what's there via a very high level interface provided by pykerberos (which is in turn still using a pretty high level interface provided by libkrb5).

**Comment by msghaleb** [comment]
I am getting the same problem too, from Ansible Tower. The command line works fine in my case, also I'm generating the ticket using the awx user on the same box - no luck on Tower.

**Comment by nitzmahone** [comment]
Tower has different issues around kerberos ccaches due to proot isolation-
contact support about ways to work around that.

**Comment by jhawkesworth** [comment]
How are you acquiring the kerberos ticket?  Does it happen automatically when you log in or are you running kinit before running your playbook.

If it happens during login, is it possible a second login is occurring during your playbook run, causing a new ticket to be stored in the cache.  That might be enough for the cache not to be there when ansible needs it.

When I made my win_domain_callback plugin I wanted to make sure I didn't hit a problem with the credential cache getting overwritten during playbook runs, so I used the pid of the ansible process and added to the end of the file name and set KRB5CCNAME env varirable, so that each playbook run would get its own credential cache.

Hope this helps,

Jon

**Comment by AseemKumar** [comment]
@jhawkesworth , I am running kinit to acquire the ticket before running the ansible command. The linux system is not joined to the domain nor has a domain computer account. 

This problem sometimes vanishes if I wait after getting the ticket and issuing the command, or one time I did kdestroy and kinit multiple times. Then, once a reboot of my linux box solved the issue and other times reboot did not help. I haven't seen a specific action from my end that is solving the problem hence unsure what exactly is going wrong.

I already did set the kinit logging and the logs never reported any error. If you can let me know any kind of debug logs that I can set which may provide more info to diagnose this, I will set it up.

**Comment by jhawkesworth** [comment]
Sorry I don't know of any logging that you can configure for kerberos.  I guess you can look in syslog and messages files.
You say issue sometimes solved by a reboot.  I had an issue a long time ago which I think had similar symptoms, which in the end was resolved by restarting nscd, but I haven't seen this since i moved off Centos 6.5.  Only other issue I recall turned out to be due to DHCP and DNS not being synchronized, which unforunately I don't have any details about how it was resolved.  DNS has to be working properly - forward and reverse lookups - for kerberos to function correctly.

**Comment by msghaleb** [comment]
I've solved it by doing this:

``` yaml
- name: generating the krb ticket
  local_action: shell bash -c "echo '{{ ansible_ssh_pass }}' | kinit {{ ansible_ssh_user }}"
  when: '"@" in ansible_ssh_user'
```

Now, using Ansible Tower it does encrypt the password but from ansible it doesn't. A proper way would be using ktutil but I can't get it to work.

**Comment by jctanner** [comment]
needs_info

**Comment by ansibot** [comment]
@AseemKumar ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@AseemKumar ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@AseemKumar closure has been requested for this issue or it has timed out waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by aaronk1** [comment]
@jhawkesworth is the win_domain_callback plugin something you're sharing?  We're having the same issue where we do multiple playbook runs and the n+1 (second or more) kinit destroys the cache the first playbook run created.  We realize that's why it's happening, just don't have a fix yet.  Assumed this was a common issue and it may be in the documentation, but haven't found anything on it.  I was going to do something like kinit -c /tmp/kinit1 user@DOMAIN.USER but figured I'd have to have logic somehow so ansible knew which cache to look in later.

**Comment by aaronk1** [comment]
@jhawkesworth Nevermind, I found it mentioned on #15310
Pointing to #14972

**Comment by nitzmahone** [comment]
@aaronk1 Have a look at 2.3 (current devel); we now do automatic ticket management with ephemeral ccaches if you specify `ansible_password` (via inventory, `-k`, whatever). Might make your life easier.

**Comment by aaronk1** [comment]
Resolved by #20416

**Comment by aaronk1** [comment]
Thanks, @nitzmahone !  Probably so.

**Comment by jhawkesworth** [comment]
@aaronk1 - please use 2.3 current devel - my win_domain callback is the past and shouldn't be used any more.

@AseemKumar Looking back at this thread, I'm almost certain the problem is clock synchronisation, having just hit this myself.  For kerberos tickets to work the ansible controller clock must be synchronised with your domain controllers.  I believe you can use ntpdate to sync against AD controllers from S2008 onwards (I hope nobody is still using a domain controller running on anything older than S2008).
